**Rudimentary Analysis:-**

In [ ]:
import numpy as np      # importing the libraries
import pandas as pd
import matplotlib.pyplot as plt             #[code snippet-1]
import seaborn as sns
# https://www.geeksforgeeks.org/bag-of-words-bow-model-in-nlp/

# https://www.geeksforgeeks.org/bag-of-words-bow-model-in-nlp/
# https://medium.com/greyatom/an-introduction-to-bag-of-words-in-nlp-ac967d43b428


In [ ]:
from google.colab import files
files=files.upload()              # uploding the csv file      [code snippet-2]

Saving train.csv.csv to train.csv.csv


In [ ]:
s=pd.read_csv('train.csv.csv') # reading the dataset by using pandas     
p=s.TripType.value_counts()    # counts of trip types
p
 #[code snippet-3]

40     174164
39      95504
37      38954
38      29565
25      27609
7       23199
8       22844
36      21990
44      20424
42      19468
24      18015
999     17590
9       16820
32      13843
5       13836
35      12501
33       9918
15       7147
3        6827
43       6383
41       5508
30       4861
34       4751
27       4613
21       4032
22       3592
6        3405
20       3116
18       2977
28       2664
26       2507
12       2108
29       2105
31       1765
19       1188
4         901
23        325
14         35
Name: TripType, dtype: int64

In [ ]:
#s_2=s[s['TripType'].isin([40,39,37,38,25,7,8,36,44,42])]
# 40    172705
# 39     94856       
# 37     38645                             #[code snippet-4]
# 38     29342
# 25     26642
# 7      23048
# 8      22630
# 36     21848
# 44     20093
# 42     19110
# since the counts these 10 data types comprises around 75% of data, so we will consider only those rows with these datatypes. These will ease our analysis.    

In [ ]:
#s_2.TripType.value_counts() # taking counts of trip types with max counts in dataset

#[code snippet-5]

In [ ]:
#s=s_2     # assigning s_2 in dataset 's'
#s.TripType.value_counts() 

#[code snippet-6]

In [ ]:
s.head() # checking the first 5 rows of our dataset

#[code snippet-7]

TripType  VisitNumber Weekday           Upc  ScanCount  \
0       999            5  Friday  6.811315e+10         -1   
1        30            7  Friday  6.053882e+10          1   
2        30            7  Friday  7.410811e+09          1   
3        26            8  Friday  2.238404e+09          2   
4        26            8  Friday  2.006614e+09          2   

   DepartmentDescription  FinelineNumber  
0     FINANCIAL SERVICES          1000.0  
1                  SHOES          8931.0  
2          PERSONAL CARE          4504.0  
3  PAINT AND ACCESSORIES          3565.0  
4  PAINT AND ACCESSORIES          1017.0

In [ ]:
s.shape # checking the size of the dataset

#[code snippet-8]

(647054, 7)

In [ ]:
s.loc[s.duplicated()].shape  # checking duplicacy in 's'

#[code snippet-9]

(7192, 7)

In [ ]:
s.loc[s.duplicated()]

#[code snippet-10]

TripType  VisitNumber Weekday           Upc  ScanCount  \
50            21           15  Friday  5.011390e+11          1   
59            42           19  Friday  7.675336e+09          1   
135           26           61  Friday  7.535309e+09          1   
461            8          225  Friday  4.093000e+03          1   
576           40          274  Friday  4.127197e+09          1   
...          ...          ...     ...           ...        ...   
646376        24       191169  Sunday  4.900003e+09          1   
646430        24       191177  Sunday  7.065200e+09          1   
646525        39       191190  Sunday  7.893138e+10          1   
646607        33       191227  Sunday  4.894150e+11          1   
646774        44       191277  Sunday  6.791453e+09          1   

       DepartmentDescription  FinelineNumber  
50        FABRICS AND CRAFTS          1309.0  
59       IMPULSE MERCHANDISE          8904.0  
135                 HARDWARE          5525.0  
461                  PRODUCE          4008.0  
576             FROZEN FOODS          9100.0  
...                      ...             ...  
646376           DSD GROCERY          9546.0  
646430         COOK AND DINE           905.0  
646525     PETS AND SUPPLIES          6210.0  
646607           CELEBRATION           306.0  
646774   IMPULSE MERCHANDISE           347.0  

[7192 rows x 7 columns]

In [ ]:
s = s.drop_duplicates()  # droping the duplicate values present in dataset
s.shape

#[code snippet-11]

(639862, 7)

In [ ]:
s.describe() # describing the numerical values present in the dataset

#[code snippet-12]

TripType    VisitNumber           Upc      ScanCount  \
count  639862.000000  639862.000000  6.369840e+05  639862.000000   
mean       58.283597   96202.936579  2.873255e+10       1.110783   
std       156.597822   55552.547291  8.262338e+10       0.702241   
min         3.000000       5.000000  8.340000e+02     -12.000000   
25%        27.000000   49321.000000  3.400001e+09       1.000000   
50%        39.000000   97103.000000  7.047049e+09       1.000000   
75%        40.000000  144367.000000  3.004101e+10       1.000000   
max       999.000000  191347.000000  9.789710e+11      71.000000   

       FinelineNumber  
count   636984.000000  
mean      3730.974387  
std       2778.910147  
min          0.000000  
25%       1404.000000  
50%       3353.000000  
75%       5501.000000  
max       9998.000000

In [ ]:
s.dtypes # describing the type of data present in 's'

#[code snippet-13]

TripType                   int64
VisitNumber                int64
Weekday                   object
Upc                      float64
ScanCount                  int64
DepartmentDescription     object
FinelineNumber           float64
dtype: object

In [ ]:

s.columns # shows columns present in dataset

#[code snippet-14]

Index(['TripType', 'VisitNumber', 'Weekday', 'Upc', 'ScanCount',
       'DepartmentDescription', 'FinelineNumber'],
      dtype='object')

In [ ]:
# Pandas "s.info()" function is used to get a concise summary of the dataframe,
# provides information about a DataFrame including the index dtype and columns, non-null values and memory usage.
# It comes really handy when doing exploratory analysis of the data.

s.info()   #[code snippet-15]

<class 'pandas.core.frame.DataFrame'>
Int64Index: 639862 entries, 0 to 647053
Data columns (total 7 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   TripType               639862 non-null  int64  
 1   VisitNumber            639862 non-null  int64  
 2   Weekday                639862 non-null  object 
 3   Upc                    636984 non-null  float64
 4   ScanCount              639862 non-null  int64  
 5   DepartmentDescription  638596 non-null  object 
 6   FinelineNumber         636984 non-null  float64
dtypes: float64(2), int64(3), object(2)
memory usage: 39.1+ MB


In [ ]:
s.isnull().sum()   #[code snippet-16]

# gives the sum of null or missing values present in each fields.
# 's.isnull()' checks null or NaN values.
# 'NaN' stands for 'Not A Number' and is one of the common ways to represent the missing value in the data.

TripType                    0
VisitNumber                 0
Weekday                     0
Upc                      2878
ScanCount                   0
DepartmentDescription    1266
FinelineNumber           2878
dtype: int64

In [ ]:
print(s.dropna(axis=0))

#[code snippet-17]

        TripType  VisitNumber Weekday           Upc  ScanCount  \
0            999            5  Friday  6.811315e+10         -1   
1             30            7  Friday  6.053882e+10          1   
2             30            7  Friday  7.410811e+09          1   
3             26            8  Friday  2.238404e+09          2   
4             26            8  Friday  2.006614e+09          2   
...          ...          ...     ...           ...        ...   
647049        39       191346  Sunday  3.239000e+10          1   
647050        39       191346  Sunday  7.874205e+09          1   
647051        39       191346  Sunday  4.072000e+03          1   
647052         8       191347  Sunday  4.190008e+09          1   
647053         8       191347  Sunday  3.800060e+09          1   

        DepartmentDescription  FinelineNumber  
0          FINANCIAL SERVICES          1000.0  
1                       SHOES          8931.0  
2               PERSONAL CARE          4504.0  
3       PAINT A

In [ ]:
s_nullrem=s.dropna(axis=0)   # dropping the null values present in 's'  

#[code snippet-18]

In [ ]:

s_nullrem.shape     # shows of the size of new dataframe with null values removed

#[code snippet-19]

(636984, 7)

In [ ]:
s_nullrem.isnull().sum()    # finding sum of null values in new dataset, to cross check our operations

#[code snippet-20]

TripType                 0
VisitNumber              0
Weekday                  0
Upc                      0
ScanCount                0
DepartmentDescription    0
FinelineNumber           0
dtype: int64

In [ ]:
s_nullrem.head() 

#[code snippet-21]

TripType  VisitNumber Weekday           Upc  ScanCount  \
0       999            5  Friday  6.811315e+10         -1   
1        30            7  Friday  6.053882e+10          1   
2        30            7  Friday  7.410811e+09          1   
3        26            8  Friday  2.238404e+09          2   
4        26            8  Friday  2.006614e+09          2   

   DepartmentDescription  FinelineNumber  
0     FINANCIAL SERVICES          1000.0  
1                  SHOES          8931.0  
2          PERSONAL CARE          4504.0  
3  PAINT AND ACCESSORIES          3565.0  
4  PAINT AND ACCESSORIES          1017.0

In [ ]:

df=s_nullrem  # assigning s_nullrem to a new variable 'df'
df

#[code snippet-22]

TripType  VisitNumber Weekday           Upc  ScanCount  \
0            999            5  Friday  6.811315e+10         -1   
1             30            7  Friday  6.053882e+10          1   
2             30            7  Friday  7.410811e+09          1   
3             26            8  Friday  2.238404e+09          2   
4             26            8  Friday  2.006614e+09          2   
...          ...          ...     ...           ...        ...   
647049        39       191346  Sunday  3.239000e+10          1   
647050        39       191346  Sunday  7.874205e+09          1   
647051        39       191346  Sunday  4.072000e+03          1   
647052         8       191347  Sunday  4.190008e+09          1   
647053         8       191347  Sunday  3.800060e+09          1   

        DepartmentDescription  FinelineNumber  
0          FINANCIAL SERVICES          1000.0  
1                       SHOES          8931.0  
2               PERSONAL CARE          4504.0  
3       PAINT AND ACCESSORIES          3565.0  
4       PAINT AND ACCESSORIES          1017.0  
...                       ...             ...  
647049           PHARMACY OTC          1118.0  
647050           FROZEN FOODS          1752.0  
647051                PRODUCE          4170.0  
647052                  DAIRY          1512.0  
647053      GROCERY DRY GOODS          3600.0  

[636984 rows x 7 columns]

In [ ]:

s_nullrem.Weekday.value_counts() # counting the number of shopping trips made on each days in a week

#[code snippet-23]

Sunday       132304
Saturday     120247
Friday        94663
Monday        81844
Tuesday       71176
Wednesday     69856
Thursday      66894
Name: Weekday, dtype: int64

In [ ]:
s_nullrem.DepartmentDescription.value_counts().head(10) #shows the top 10 department by number of rows

#[code snippet-24]

GROCERY DRY GOODS           70033
DSD GROCERY                 67682
PRODUCE                     50826
DAIRY                       43599
PERSONAL CARE               41885
IMPULSE MERCHANDISE         28458
HOUSEHOLD CHEMICALS/SUPP    24804
PHARMACY OTC                23243
FROZEN FOODS                20984
HOUSEHOLD PAPER GOODS       16211
Name: DepartmentDescription, dtype: int64

**Observations[code snippet:1-24]:-**

1- Basic informations about the dataset is obtained.

2- Null and Duplicate values present in data set has been removed.


**Data preprocessing and Training of the dataset:-**

In [ ]:
dept=['HEALTH AND BEAUTY AIDS','OTHER DEPARTMENTS','SEASONAL','CONCEPT STORES',
      'LARGE HOUSEHOLD GOODS','OPTICAL - LENSES','PHARMACY RX','CAMERAS AND SUPPLIES',
      'MENSWEAR','1-HR PHOTO','SHEER HOSIERY','FURNITURE','OPTICAL - FRAMES',
      'PLUS AND MATERNITY','PLAYERS AND ELECTRONICS','LADIES SOCKS','SWIMWEAR/OUTERWEAR',
      'BOOKS AND MAGAZINES','ACCESSORIES','PAINT AND ACCESSORIES',
     'WIRELESS','SEAFOOD','MEDIA AND GAMING','BRAS & SHAPEWEAR','BEDDING','SLEEPWEAR/FOUNDATIONS',
      'JEWELRY AND SUNGLASSES','HORTICULTURE AND ACCESS','ELECTRONICS','LIQUOR,WINE,BEER','BOYS WEAR',
      'HOME DECOR','GIRLS WEAR, 4-6X  AND 7-14','BATH AND SHOWER','HARDWARE','FABRICS AND CRAFTS','SPORTING GOODS',
      'LAWN AND GARDEN','AUTOMOTIVE','INFANT APPAREL','SHOES','TOYS','HOME MANAGEMENT','OFFICE SUPPLIES','BAKERY',
      'COOK AND DINE','CELEBRATION','LADIESWEAR','PRE PACKED DELI','MEAT - FRESH & FROZEN','PETS AND SUPPLIES',
      'CANDY, TOBACCO, COOKIES','SERVICE DELI','INFANT CONSUMABLE HARDLINES','FINANCIAL SERVICES','MENS WEAR',
      'BEAUTY','COMM BREAD','HOUSEHOLD PAPER GOODS','FROZEN FOODS','PHARMACY OTC','HOUSEHOLD CHEMICALS/SUPP','IMPULSE MERCHANDISE',
      'PERSONAL CARE','DAIRY','PRODUCE','DSD GROCERY','GROCERY DRY GOODS']
      
      # forming a list of unique departments in the column 'DepartmentDescription', by arranging the departments in ascending order to their shopping counts.
      # this list will help in 'encoding' of the dataset.

      #[code snippet-25]

In [ ]:

days=['Thursday','Wednesday','Tuesday','Monday','Friday','Saturday','Sunday']

  # forming a list of days in the column 'Weekday', by arranging the days in ascending order to their shopping counts.
   # this list will help in 'encoding' of the dataset.

   #[code snippet-26]

In [ ]:
#x=df[["Weekday","ScanCount","DepartmentDescription"]]
x=df[["Weekday","DepartmentDescription"]] # assigning independent columns used for analysis in variable 'x' 
x

#[code snippet-27]

Weekday  DepartmentDescription
0       Friday     FINANCIAL SERVICES
1       Friday                  SHOES
2       Friday          PERSONAL CARE
3       Friday  PAINT AND ACCESSORIES
4       Friday  PAINT AND ACCESSORIES
...        ...                    ...
647049  Sunday           PHARMACY OTC
647050  Sunday           FROZEN FOODS
647051  Sunday                PRODUCE
647052  Sunday                  DAIRY
647053  Sunday      GROCERY DRY GOODS

[636984 rows x 2 columns]

In [ ]:
y=df.iloc[:,0]   # assigning dependent columns used for analysis in variable 'y' 
y

#[code snippet-28]

0         999
1          30
2          30
3          26
4          26
         ... 
647049     39
647050     39
647051     39
647052      8
647053      8
Name: TripType, Length: 636984, dtype: int64

In [ ]:
y.value_counts() # checking counts of 'y'

#[code snippet-29]

40     172705
39      94856
37      38645
38      29342
25      26642
7       23048
8       22630
36      21848
44      20093
42      19110
24      17663
999     16869
9       16555
32      13703
35      12339
5       11181
33       9850
15       7047
3        6814
43       6270
41       5404
30       4805
34       4684
27       4533
21       3957
22       3553
6        3369
20       3090
18       2931
28       2613
26       2477
12       2092
29       2090
31       1752
19       1183
4         889
23        323
14         29
Name: TripType, dtype: int64

In [ ]:
seed=7     #[code snippet-30]

In [ ]:

# dividing the dataset into train-test, by using 'train_test_split' sub module present in model_selection module of 'skleartn' library
  
testsize=0.33
from sklearn.model_selection import train_test_split 
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=testsize,random_state=seed)

#[code snippet-31]

In [ ]:
x_train,x_train.shape # checking train data and its shape.

#[code snippet-32]

(         Weekday     DepartmentDescription
 36745   Saturday                   PRODUCE
 336532    Sunday   CANDY, TOBACCO, COOKIES
 518254    Monday         GROCERY DRY GOODS
 286405  Thursday             PERSONAL CARE
 55245     Sunday  HOUSEHOLD CHEMICALS/SUPP
 ...          ...                       ...
 334129    Sunday       IMPULSE MERCHANDISE
 543617   Tuesday              FROZEN FOODS
 583134    Friday                 MENS WEAR
 192841    Sunday               DSD GROCERY
 595276    Friday              PHARMACY OTC
 
 [426779 rows x 2 columns], (426779, 2))

In [ ]:
x_test,x_test.shape   # checking test data and its shape.

#[code snippet-33]

(         Weekday        DepartmentDescription
 134473  Thursday                  DSD GROCERY
 279014  Thursday     HOUSEHOLD CHEMICALS/SUPP
 453751    Friday                 PHARMACY OTC
 30662   Saturday  INFANT CONSUMABLE HARDLINES
 75547     Monday                      PRODUCE
 ...          ...                          ...
 494772    Sunday                      SEAFOOD
 188876    Sunday      HORTICULTURE AND ACCESS
 21028   Saturday            PETS AND SUPPLIES
 16137     Friday            GROCERY DRY GOODS
 84080     Monday                   COMM BREAD
 
 [210205 rows x 2 columns], (210205, 2))

**Observations[code snippet:25-33]:-**

1- Dependent and independent variables i.e x and y have been formed.

2- The dataset has been split for training and testing.

3- The features present in the dataset are prepared for encoding.

**Encoding the dataset(1st method):-** In this we will encode the column 'Weekday' by using 'One hot encoding' and column 'DepartmentDescription' will be encoded by using 'Ordinal Encoding'.
 

In [ ]:
from sklearn.preprocessing import OrdinalEncoder # importing the preprocessing libraries/modules
#[code snippet-34]

In [ ]:
from sklearn.preprocessing import OneHotEncoder
#[code snippet-35]

In [ ]:
from sklearn.compose import ColumnTransformer  # importing the column transformer
#[code snippet-36]

In [ ]:

#transformer=ColumnTransformer(transformers=[
 #   ('tnf1',OrdinalEncoder(categories=[dept]),['DepartmentDescription']),
  #  ('tnf2',OneHotEncoder(sparse=False,drop='first'),['Weekday'])
#],remainder='passthrough')   

In [ ]:
transformer=ColumnTransformer(transformers=[
    ('tnf1',OneHotEncoder(sparse=False,drop='first'),['DepartmentDescription']),
    ('tnf2',OneHotEncoder(sparse=False,drop='first'),['Weekday'])
],remainder='passthrough')                                         

# applying column transform in columns 'DepartmentDescription' and 'Weekday', by using their respective encoding techniques.
#[code snippet-37]

In [ ]:
x1_train=pd.DataFrame(transformer.fit_transform(x_train))  # fitting the transform in the train data after encoding, and converting it into a dataframe.
x1_train
#[code snippet-38]


0    1    2    3    4    5    6    7    8    9   ...   63   64   65  \
0       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
426774  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
426775  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
426776  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
426777  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
426778  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

         66   67   68   69   70   71   72  
0       0.0  0.0  1.0  0.0  0.0  0.0  0.0  
1       0.0  0.0  0.0  1.0  0.0  0.0  0.0  
2       0.0  1.0  0.0  0.0  0.0  0.0  0.0  
3       0.0  0.0  0.0  0.0  1.0  0.0  0.0  
4       0.0  0.0  0.0  1.0  0.0  0.0  0.0  
...     ...  ...  ...  ...  ...  ...  ...  
426774  0.0  0.0  0.0  1.0  0.0  0.0  0.0  
426775  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
426776  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
426777  0.0  0.0  0.0  1.0  0.0  0.0  0.0  
426778  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[426779 rows x 73 columns]

In [ ]:
y_train #[code snippet-39]

36745     37
336532    24
518254    39
286405    37
55245     39
          ..
334129    38
543617     5
583134    25
192841    43
595276    39
Name: TripType, Length: 426779, dtype: int64

In [ ]:
type(y_train)     #[code snippet-40]

pandas.core.series.Series

In [ ]:
y1_train=pd.DataFrame(y_train)  #converting the y_train into dataframe
y1_train
#[code snippet-41]

TripType
36745         37
336532        24
518254        39
286405        37
55245         39
...          ...
334129        38
543617         5
583134        25
192841        43
595276        39

[426779 rows x 1 columns]

In [ ]:
type(y1_train)      #[code snippet-42]

pandas.core.frame.DataFrame

In [ ]:
x1_test=pd.DataFrame(transformer.transform(x_test))    # fitting the transform in the test data after encoding, and converting it into a dataframe.
x1_test
#[code snippet-43]

0    1    2    3    4    5    6    7    8    9   ...   63   64   65  \
0       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
2       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
3       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
210200  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
210201  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
210202  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
210203  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
210204  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

         66   67   68   69   70   71   72  
0       0.0  0.0  0.0  0.0  1.0  0.0  0.0  
1       0.0  0.0  0.0  0.0  1.0  0.0  0.0  
2       0.0  0.0  0.0  0.0  0.0  0.0  0.0  
3       0.0  0.0  1.0  0.0  0.0  0.0  0.0  
4       0.0  1.0  0.0  0.0  0.0  0.0  0.0  
...     ...  ...  ...  ...  ...  ...  ...  
210200  0.0  0.0  0.0  1.0  0.0  0.0  0.0  
210201  0.0  0.0  0.0  1.0  0.0  0.0  0.0  
210202  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
210203  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
210204  0.0  1.0  0.0  0.0  0.0  0.0  0.0  

[210205 rows x 73 columns]

**Observations[code snippet:34-43]:-**

1- Categorical columns 'Weekday' and 'DepartmentDescription' are encoded with 'One hot encoding' 

2- Column Transformer has been used to apply transforms to selected colums present in dataset after encoding.

3- Dataset is prepared for the implementation of Maching learning algorithms.

**Implementation of Machine Learning Algorithms:-**

**Model 1 - Random Forest Classifier**

In [ ]:
from sklearn.ensemble import RandomForestClassifier # importing RandomForestClassifier from sklearn
#[code snippet-44]

In [ ]:
rfg=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',   # initializing an unstance for RandomForestClassifier
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False) 
#[code snippet-45]

In [ ]:
#loss=[]
#for i in [7,8,9,10,11,12]:
#   rfg=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',   # initializing an unstance for RandomForestClassifier
#                       max_depth=None, max_features='auto', max_leaf_nodes=None,
#                       min_impurity_decrease=0.0,
#                       min_samples_leaf=1, min_samples_split=2,
#                       min_weight_fraction_leaf=0.0, n_estimators=i,
#                       n_jobs=None, oob_score=False, random_state=None,
#                       verbose=0, warm_start=False)
#   rfg.fit(x1_train,y1_train)

In [ ]:
rfg.fit(x1_train,y1_train)  # fitting the RandomForestClassifier model in train data

#[code snippet-46]

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  """Entry point for launching an IPython kernel.


RandomForestClassifier(n_estimators=10)

In [ ]:
y_pred=rfg.predict(x1_test) # predicting the result of training the model, by using test data 
#[code snippet-47]

In [ ]:
y_pred    #[code snippet-48]

array([40, 40,  5, ..., 40, 40, 40])

In [ ]:
y_test   #[code snippet-49]

134473     39
279014     37
453751      4
30662      32
75547       8
         ... 
494772     12
188876    999
21028      34
16137      40
84080      40
Name: TripType, Length: 210205, dtype: int64

In [ ]:
y_rfg_pred_train=rfg.predict(x1_train)  # applying Random forest model in train data to check robustness of model
y_rfg_pred_train
#[code snippet-50]

array([40, 40, 40, ..., 25, 40,  5])

**Model Performance Analysis:-**

In [ ]:

from sklearn.metrics import confusion_matrix   # Model Performance Analysis, by using confusion matrix

# confusion matrix:- it is a better way to evaluate the performance of a classifier.
# The general idea is to count the number of times instances of class A are classified as class B.
# Each row in a confusion matrix represents an actual class, while each column represents a predicted class or vice versa.

#[code snippet-51]


In [ ]:
confusion_matrix(y_test,y_pred)
pd.crosstab(y_test, y_pred, rownames=['True'], colnames=['Predicted'],margins=True)

# crosstab:- This method is used to compute a simple cross-tabulation of two (or more) factors.
# By default, computes a frequency table of the factors unless an array of values and an aggregation function are passed.
 
 # index(y_test) : array-like, Series, or list of arrays/Series, Values to group by in the rows.
 # columns(y_pred) : array-like, Series, or list of arrays/Series, Values to group by in the columns.
 # rownames : sequence, default None, If passed, must match number of row arrays passed.
 # colnames : sequence, default None, If passed, must match number of column arrays passed.
 # margins : bool, default False, Add row/column margins (subtotals).
 # Row indicates the actual values of data and columns indicate the predicted data.

 #[code snippet-52]

Predicted     3     5    6     7     9    15    18   19    20    21  ...   34  \
True                                                                 ...        
3          1833     9    2     8     3     1     3    1     8     6  ...    0   
4             1   140    0     1     0     1     0    0     1     0  ...    0   
5            15  1438   15    32    17    32     6    1     8    51  ...    6   
6             6     6  206    16     0    13     0    0     2     9  ...    5   
7            23    43    9  1070    10    13     5    0    10    19  ...    4   
8             0   135   15    65     7    26     8    2    16    48  ...   50   
9            19    43    2    22   177   207   159   60   305   333  ...    1   
12            0     4    3     7     6    11    22    1     3     8  ...    1   
14            1     0    0     0     0     1     0    0     0     2  ...    0   
15           16    14    7    37    10   765    20    0     1    87  ...    0   
18           11    10    1     5     1    68   364    1     6    28  ...    0   
19           18     5    1     0     6     3     2   44     3    19  ...    0   
20           16    21    2     7    11     4     5    0   544    11  ...    2   
21            8    10    0     4     2    44    15    0     0   811  ...    0   
22           33    20    1     7    27     9     7   59    11    21  ...    1   
23            8     4    0     1    22     0     1    1     0     0  ...    0   
24           67    69    9    33    67    60    20    6    26   150  ...    7   
25           56   142   15    62    34    82    42    2    27   138  ...   12   
26            9    14    2     6   108     4     1    0    16    13  ...    1   
27           10    19    5     3    14    20     7    0     5    17  ...    6   
28           12     6    2     5     4    10     7    0     7     9  ...    0   
29           12    12    1     5     3     2   141    1    12     6  ...    1   
30           22    16    1     6    43    17    10    2     6    18  ...    1   
31            6    13    5     5     2     2     0    4     4     5  ...    0   
32           27    58    7    25     8    70    58    0    14    39  ...    1   
33           15    32    3    14    11    15     5    2     8    35  ...    4   
34           13    21    3    11    11     5     4    1    14     7  ...  194   
35           20    28    8    34     5    11    10    0    11    33  ...    5   
36           45   370   11    22    19    70    28    0    34    68  ...   10   
37           41    89   39   186    14    34    31    1    22    31  ...    9   
38           41    75   10    67    12    32    23    0    11    41  ...   11   
39          132   997  167   417   105   198   149   17   145   258  ...  125   
40          174   836  204   479   112   261   189   18   129   360  ...  146   
41           12    17    8     4    42    47    41    3    36    45  ...    4   
42           25    92   10    33    88   366   226    7   122   441  ...    9   
43           20    81    5    29    26    69    38    3    41   139  ...    5   
44           46   176   34    34    71   180    83    7    81   271  ...   19   
999         767   125    7    52    91    67   109   55   117   110  ...    6   
All        3580  5190  820  2814  1189  2820  1839  299  1806  3687  ...  646   

Predicted    36    37  38    39      40   42  44  999     All  
True                                                           
3             4     2   0   209     137    0   0    0    2297  
4             3     3   0    12     133    0   0    0     303  
5           109    40   0   151    1599    6   0    2    3743  
6             8    11   0   271     525    0   0    0    1126  
7            31   256   0   216    5795    4   0    1    7674  
8           412   163   0   717    5434    1   0    1    7386  
9            43    16   0   341     762   30   0   39    5333  
12           10     4   0    19     467    1   0    1     701  
14            0     0   0     0       

In [ ]:
from sklearn.metrics import classification_report   
                                                            
print(classification_report(y_test,y_pred)) 

# Classification Report:- As the name suggests, it is the report which explains everything about the classification.
# This is the summary of the quality of classification made by the constructed ML model

# Precision Score:- it is the accuracy of the positive predictions.
               # TP - True positive; FP - False positive
               # precision = TP/(TP+FP)                                                         
# Recall Score:- fraction of the positives correctly identified
               # FN - False Negative
               # Recall = TP/(TP+FN)
# F1 Score:- takes both precision and recall into account; 
               # F1 Score = 2(precision*recall)/(precision+recall)
# Support: It is the total entries of each class in the actual dataset. It is simply the sum of rows for every class.  

#[code snippet-53]

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

           3       0.51      0.80      0.62      2297
           4       0.00      0.00      0.00       303
           5       0.28      0.38      0.32      3743
           6       0.25      0.18      0.21      1126
           7       0.38      0.14      0.20      7674
           8       0.00      0.00      0.00      7386
           9       0.15      0.03      0.05      5333
          12       0.00      0.00      0.00       701
          14       0.00      0.00      0.00         6
          15       0.27      0.33      0.30      2349
          18       0.20      0.37      0.26       988
          19       0.15      0.11      0.13       388
          20       0.30      0.54      0.39      1015
          21       0.22      0.63      0.33      1280
          22       0.27      0.31      0.29      1191
          23       0.00      0.00      0.00       109
          24       0.31      0.44      0.37      5783
          25       0.41    

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
from sklearn.metrics import roc_auc_score
y_score_1=rfg.predict_proba(x1_test)

# roc_auc_score is used for evaluating the performance of the classification model,it also helps in determining the robustness of the model.
#[code snippet-54]

In [ ]:
roc_auc_score(y_test,y_score_1,multi_class='ovr')
#[code snippet-55]

0.8016803486683756

In [ ]:
confusion_matrix(y_train,y_rfg_pred_train)  # confusion matrix for train data
pd.crosstab(y_train, y_rfg_pred_train, rownames=['True'], colnames=['Predicted'],margins=True)
#[code snippet-56]

Predicted     3      5     6     7     9    15    18   19    20    21  ...  \
True                                                                   ...   
3          3606     24     2     5     1    16     3    1    12    13  ...   
4             4    240     0     2     0     2     0    0     2     2  ...   
5            35   2790    18    55    34    51     9    0    30    93  ...   
6             8     31   474    32     4    17     3    0     3    16  ...   
7            42     65    30  2113    18    31    14    3    21    50  ...   
8             0    308    21   127    28    75    18    2    40    76  ...   
9            42    109     4    41   410   511   345   98   577   727  ...   
12            7     13     6     7     1    13    42    1     1    11  ...   
14            1      0     0     0     0     2     0    0     0     9  ...   
15           34     38     8    77    13  1506    23    2     4   214  ...   
18           21     28     1     8     6   109   704    1    11    63  ...   
19           29     22     3     2    13     4    13  116     5    32  ...   
20           32     32     4    14    26     5     9    3  1084    18  ...   
21           18     33     3     6    12    80    27    2     5  1716  ...   
22           62     28     5    22    62    28    12   93    15    29  ...   
23            8      2     0     1    37     1     0    0     0     7  ...   
24          121    164    17    58   136   118    54    8    44   339  ...   
25          142    222    24   120    87   158    90   10    48   248  ...   
26           20     34     2     8   181    10     7    0    23    41  ...   
27           24     49    11    16    11    37    18    0    16    37  ...   
28           28     17     4     6     8     8    14    3     8    21  ...   
29           25     32     4     5    11     7   295    2    18    17  ...   
30           34     59     5    23    91    30     6    0    19    76  ...   
31           21     18    12     5     7     7     5    3     5     9  ...   
32           64    121     7    57    23   128   109    1    29   101  ...   
33           33     64     6    37    23    22    20    1    23    41  ...   
34           21     49     3    22    11     8     4    1    36    22  ...   
35           48     58    14    86     7    33    14    2    16    53  ...   
36           78    713    22    61    49   130    47    2    54   165  ...   
37           80    186    54   282    21    52    48    4    36    86  ...   
38           67    135    20   132    18    58    30    1    45    98  ...   
39          307   1960   395   887   184   379   243   29   287   541  ...   
40          349   1779   405  1003   189   566   381   31   279   760  ...   
41           27     47    11    25    88    72    84   18    68   103  ...   
42           94    177    24    54   199   758   502   23   254   945  ...   
43           72    134    13    36    50   151    73    4    88   285  ...   
44           83    320    61    80   141   432   187   13   128   632  ...   
999        1523    237    10    66   163   170   177   96   225   262  ...   
All        7210  10338  1703  5581  2363  5785  3630  574  3559  7958  ...   

Predicted    34     36    37  38     39      40   42  44  999     All  
True                                                                   
3             0     23     3   0    407     272    0   0    7    4517  
4             0      8    11   0     21     276    2   0    0     586  
5            12    198    81   0    339    3268    4   0    7    7438  
6             2     23    14   0    484    1051    2   0    1    2243  
7             6     64   539   0    464   11586    2   0    4   15374  
8           106    832   331   0   1486   11204    7   0    2   15244  
9             6    117    28   0    711    1533   77   0  108   11222  
12            1      9     5   0     29     966    0   0    0    1391  
14            0      0     0   0      4       5    0   0    0      23  
15       

In [ ]:
from sklearn.metrics import classification_report   # classification report for train data
                                                            
print(classification_report(y_train,y_rfg_pred_train))
#[code snippet-57]

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

           3       0.50      0.80      0.61      4517
           4       0.00      0.00      0.00       586
           5       0.27      0.38      0.31      7438
           6       0.28      0.21      0.24      2243
           7       0.38      0.14      0.20     15374
           8       0.00      0.00      0.00     15244
           9       0.17      0.04      0.06     11222
          12       0.00      0.00      0.00      1391
          14       0.00      0.00      0.00        23
          15       0.26      0.32      0.29      4698
          18       0.19      0.36      0.25      1943
          19       0.20      0.15      0.17       795
          20       0.30      0.52      0.38      2075
          21       0.22      0.64      0.32      2677
          22       0.25      0.31      0.28      2362
          23       0.00      0.00      0.00       214
          24       0.32      0.45      0.38     11880
          25       0.42    

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
y_score_1t=rfg.predict_proba(x1_train)    #[code snippet-58]

In [ ]:
roc_auc_score(y_train,y_score_1t,multi_class='ovr') # roc_auc_score for train data
#[code snippet-59]

0.8294380239481565

**Hyper Parameter Tuning:-**

**Randomized Search CV**

In [ ]:
n_estimators=[100,200,400,500]   # Number of trees in random forest

max_features=['sqrt','log2',None]   # Number of features to consider at every split

max_depth=[2,8,12,16,24,None]       # Maximum number of levels in tree

max_samples=[0.5,0.75,1.0]           # Number of samples

#[code snippet-60]

In [ ]:

param_grid={'n_estimators':n_estimators, 'max_features': max_features, 'max_depth':max_depth, 'max_samples':max_samples, 'random_state':[seed],
            'criterion':['gini','entropy'], 'min_samples_leaf':[20,40,60,80,100,300,500]}   # parameters value dictionary for random search cv.

#[code snippet-61]            

In [ ]:
from sklearn.model_selection import RandomizedSearchCV    # importing RandomizedSearchCV
rf_grid1=RandomizedSearchCV(estimator=rfg, param_distributions=param_grid, cv=5, verbose=2, n_jobs=-1)

#[code snippet-62]

In [ ]:
rf_grid1.fit(x1_train,y1_train)    # fitting ramdom search cv in train data

#[code snippet-63]

Fitting 5 folds for each of 10 candidates, totalling 50 fits


/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_search.py:926: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  self.best_estimator_.fit(X, y, **fit_params)


RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(n_estimators=10),
                   n_jobs=-1,
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'max_depth': [2, 8, 12, 16, 24, None],
                                        'max_features': ['sqrt', 'log2', None],
                                        'max_samples': [0.5, 0.75, 1.0],
                                        'min_samples_leaf': [20, 40, 60, 80,
                                                             100, 300, 500],
                                        'n_estimators': [100, 200, 400, 500],
                                        'random_state': [7]},
                   verbose=2)

In [ ]:
rf_grid1.best_params_         # shows the best parameters suitable for good tuning of random forest model

#[code snippet-64]

{'random_state': 7,
 'n_estimators': 500,
 'min_samples_leaf': 60,
 'max_samples': 0.75,
 'max_features': 'sqrt',
 'max_depth': 24,
 'criterion': 'gini'}

In [ ]:
rf_grid1.best_score_        # shows the best score obtained after tuning of hyper parameters

#[code snippet-65]

0.3165033902713479

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix,precision_score,f1_score,recall_score

In [ ]:
n_estimators=[100,300]   # Number of trees in random forest
   
max_depth=[8,16,None]     # max deapth of tree 

results_1=[]             #using custom loop for hyperparameter tuning

for i in n_estimators:
  for k in max_depth:
    model_1= RandomForestClassifier(n_estimators=i,max_depth=k,random_state=seed)
    model_1.fit(x1_train,y1_train)
    preds=model_1.predict(x1_test)
    results_1.append({'Accuracy':round(accuracy_score(y_test,preds),5),'Precision':round(precision_score(y_test,preds,average='macro'),5),'Recall':round(recall_score(y_test,preds,average='macro'),5),'F1-score':round(f1_score(y_test,preds,average='macro'),5),'n_estimators':i,'max_depth':k})
results_1=pd.DataFrame(results_1)
results_1=results_1.sort_values(by='Accuracy',ascending=False)
print(results_1)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  # Remove the CWD from sys.path while we load stuff.
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:10: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  # Remove the CWD from sys.path while we load stuff.
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in l

   Accuracy  Precision   Recall  F1-score  n_estimators  max_depth
5   0.33656    0.20771  0.23595   0.19562           300        NaN
2   0.33653    0.20679  0.23484   0.19479           100        NaN
4   0.30488    0.08405  0.07973   0.06553           300       16.0
1   0.30433    0.07115  0.07917   0.06407           100       16.0
3   0.29302    0.04436  0.06081   0.04531           300        8.0
0   0.29254    0.04477  0.05965   0.04417           100        8.0


In [ ]:
n_estimators=[100,300]   # Number of trees in random forest
max_features=['sqrt','log2',None]   
max_depth=[8,16,None]      

results_1=[]

for i in n_estimators:
  for j in max_depth:
    for k in max_features: 
      model_1= RandomForestClassifier(n_estimators=i,max_depth=j,max_features=k,random_state=seed)
      model_1.fit(x1_train,y1_train)
      preds=model_1.predict(x1_test)
      results_1.append({'Accuracy':round(accuracy_score(y_test,preds),5),'Precision':round(precision_score(y_test,preds,average='macro'),5),'Recall':round(recall_score(y_test,preds,average='macro'),5),'F1-score':round(f1_score(y_test,preds,average='macro'),5),'n_estimators':i,'max_depth':j,'max_features':k})
results_1=pd.DataFrame(results_1)
results_1=results_1.sort_values(by='Accuracy',ascending=False)
print(results_1)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:11: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  # This is added back by InteractiveShellApp.init_path()
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 

    Accuracy  Precision   Recall  F1-score  n_estimators  max_depth  \
16   0.33657    0.20773  0.23595   0.19563           300        NaN   
17   0.33656    0.20773  0.23595   0.19563           300        NaN   
15   0.33656    0.20771  0.23595   0.19562           300        NaN   
6    0.33653    0.20679  0.23484   0.19479           100        NaN   
7    0.33653    0.20679  0.23484   0.19479           100        NaN   
8    0.33653    0.20679  0.23484   0.19479           100        NaN   
12   0.30488    0.08405  0.07973   0.06553           300       16.0   
4    0.30444    0.07736  0.08889   0.07201           100       16.0   
3    0.30433    0.07115  0.07917   0.06407           100       16.0   
13   0.30431    0.07738  0.08556   0.07015           300       16.0   
5    0.29938    0.07392  0.07739   0.06426           100       16.0   
14   0.29938    0.07392  0.07739   0.06426           300       16.0   
9    0.29302    0.04436  0.06081   0.04531           300        8.0   
0    0

**Observations[code snippet:44-65]:-**

1- Random forest classifier is applied to the dataset for the classification of Trip Types.

2- Accuracy of 34%, test roc_auc score is 80% and train roc_auc score is 82%.

3- Randomized search CV and custom loop is used for Hyper parameter tuning.

4- Best score obtained after hyper parameter tuning is 33.7%.

**Model 2 - Decision Tree classifier**

In [ ]:
from sklearn.tree import DecisionTreeClassifier    # importing DecisionTreeClassifier 

#[code snippet-66]

In [ ]:
dtc=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,  # initializing an instance for DecisionTreeClassifier  
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0,
                       random_state=None, splitter='best')

#[code snippet-67]

In [ ]:
dtc.fit(x1_train,y1_train) # fitting the decision tree model in train data.

#[code snippet-68]

DecisionTreeClassifier()

In [ ]:
y_dtc_pred=dtc.predict(x1_test) # after traing the train data, predicting the result in test data. 

#[code snippet-69]

In [ ]:
y_dtc_pred      
#[code snippet-70]

array([40, 40,  5, ..., 40, 40, 40])

In [ ]:
y_dtc_pred_train=dtc.predict(x1_train)     # applying  model in train data to check robustness of model
y_dtc_pred_train
#[code snippet-71]

array([40, 40, 40, ..., 25, 40,  5])

In [ ]:
confusion_matrix(y_test,y_dtc_pred)  # using confusion matrix for evaluating the performance of model
pd.crosstab(y_test, y_dtc_pred, rownames=['True'], colnames=['Predicted'],margins=True) 
#[code snippet-72]

Predicted     3     5    6     7     9    15    18   19    20    21  ...  \
True                                                                 ...   
3          1833     9    2     8     5     1     3    1     8     6  ...   
4             1   140    0     1     0     1     0    0     1     0  ...   
5            15  1438   15    32    19    32     6    1     8    51  ...   
6             6     6  206    16     0    13     0    0     2     9  ...   
7            23    43    9  1070    14    13     5    0    10    19  ...   
8             0   135   15    65    10    26     8    2    16    48  ...   
9            19    43    2    22   199   207   159   60   305   333  ...   
12            0     4    3     7     7    11    22    1     3     8  ...   
14            1     0    0     0     0     1     0    0     0     2  ...   
15           16    14    7    37    15   765    20    0     1    87  ...   
18           11    10    1     5     2    68   364    1     6    28  ...   
19           18     5    1     0     2     3     2   44     3    19  ...   
20           16    21    2     7    10     4     5    0   544    11  ...   
21            8    10    0     4     2    44    15    0     0   811  ...   
22           33    20    1     7    22     9     7   59    11    21  ...   
23            8     4    0     1     2     0     1    1     0     0  ...   
24           67    69    9    33    75    60    20    6    26   150  ...   
25           56   142   15    62    80    82    42    2    27   138  ...   
26            9    14    2     6   108     4     1    0    16    13  ...   
27           10    19    5     3    15    20     7    0     5    17  ...   
28           12     6    2     5     4    10     7    0     7     9  ...   
29           12    12    1     5     3     2   141    1    12     6  ...   
30           22    16    1     6    95    17    10    2     6    18  ...   
31            6    13    5     5     1     2     0    4     4     5  ...   
32           27    58    7    25    26    70    58    0    14    39  ...   
33           15    32    3    14    12    15     5    2     8    35  ...   
34           13    21    3    11    11     5     4    1    14     7  ...   
35           20    28    8    34     5    11    10    0    11    33  ...   
36           45   370   11    22    18    70    28    0    34    68  ...   
37           41    89   39   186    15    34    31    1    22    31  ...   
38           41    75   10    67    15    32    23    0    11    41  ...   
39          132   997  167   417   126   198   149   17   145   258  ...   
40          174   836  204   479   117   261   189   18   129   360  ...   
41           12    17    8     4    56    47    41    3    36    45  ...   
42           25    92   10    33   126   366   226    7   122   441  ...   
43           20    81    5    29    33    69    38    3    41   139  ...   
44           46   176   34    34    98   180    83    7    81   271  ...   
999         767   125    7    52   102    67   109   55   117   110  ...   
All        3580  5190  820  2814  1450  2820  1839  299  1806  3687  ...   

Predicted    32   34    36    37  38    39      40   42  999     All  
True                                                                  
3             6    0     4     2   0   205     141    0    0    2297  
4             0    0     3     3   0    11     134    0    0     303  
5            10    6   109    21   0   140    1628    5    2    3743  
6             2    5     8     7   0   186     614    0    0    1126  
7            18    4    31   114   0   176    5979    1    1    7674  
8           132   50   412    85   0   647    5582    0    1    7386  
9           129    1    43    10   0   311     779   37   39    5333  
12            1    1    10     2   0    14     474    1    1     701  
14            0    0     0     0   0     0       2    0    0       6  
15           37    0    20    21   0   129    1037    4    0    2349  
18           19    0    14     

In [ ]:
from sklearn.metrics import classification_report   # implementing classification report
                                                            
print(classification_report(y_test,y_dtc_pred)) 
#[code snippet-73]

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

           3       0.51      0.80      0.62      2297
           4       0.00      0.00      0.00       303
           5       0.28      0.38      0.32      3743
           6       0.25      0.18      0.21      1126
           7       0.38      0.14      0.20      7674
           8       0.00      0.00      0.00      7386
           9       0.14      0.04      0.06      5333
          12       0.00      0.00      0.00       701
          14       0.00      0.00      0.00         6
          15       0.27      0.33      0.30      2349
          18       0.20      0.37      0.26       988
          19       0.15      0.11      0.13       388
          20       0.30      0.54      0.39      1015
          21       0.22      0.63      0.33      1280
          22       0.27      0.32      0.29      1191
          23       0.32      0.12      0.17       109
          24       0.31      0.44      0.37      5783
          25       0.41    

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
from sklearn.metrics import roc_auc_score      # calculating the roc_auc_score for decision tree model
y_score=dtc.predict_proba(x1_test)
#[code snippet-74]

In [ ]:
roc_auc_score(y_test,y_score,multi_class='ovr')
#[code snippet-75]

0.8025383596333668

In [ ]:
confusion_matrix(y_train,y_dtc_pred_train)           # confusion matrix for train data
pd.crosstab(y_train, y_dtc_pred_train, rownames=['True'], colnames=['Predicted'],margins=True)
#[code snippet-76]

Predicted     3      5     6     7     9    15    18   19    20    21  ...  \
True                                                                   ...   
3          3606     24     2     5     4    16     3    1    12    13  ...   
4             4    240     0     2     0     2     0    0     2     2  ...   
5            35   2791    18    55    41    51     9    0    30    93  ...   
6             8     31   474    32     5    17     3    0     3    16  ...   
7            42     65    30  2113    19    31    14    3    21    50  ...   
8             0    308    21   127    32    75    18    2    40    76  ...   
9            42    109     4    41   495   511   345   98   577   727  ...   
12            7     13     6     7     2    13    42    1     1    11  ...   
14            1      0     0     0     0     2     0    0     0     9  ...   
15           34     38     8    77    23  1506    23    2     4   214  ...   
18           21     28     1     8     9   109   704    1    11    63  ...   
19           29     22     3     2    12     4    13  116     5    32  ...   
20           32     32     4    14    25     5     9    3  1084    18  ...   
21           18     33     3     6    16    80    27    2     5  1716  ...   
22           62     28     5    22    38    28    12   93    15    29  ...   
23            8      2     0     1     5     1     0    0     0     7  ...   
24          121    164    17    58   155   118    54    8    44   339  ...   
25          142    222    24   120   134   158    90   10    48   248  ...   
26           20     34     2     8   182    10     7    0    23    41  ...   
27           24     49    11    16    14    37    18    0    16    37  ...   
28           28     17     4     6    10     8    14    3     8    21  ...   
29           25     32     4     5    13     7   295    2    18    17  ...   
30           34     59     5    23   198    30     6    0    19    76  ...   
31           21     18    12     5     3     7     5    3     5     9  ...   
32           64    121     7    57    48   128   109    1    29   101  ...   
33           33     64     6    37    25    22    20    1    23    41  ...   
34           21     49     3    22    13     8     4    1    36    22  ...   
35           48     58    14    86     8    33    14    2    16    53  ...   
36           78    713    22    61    61   130    47    2    54   165  ...   
37           80    186    54   282    24    52    48    4    36    86  ...   
38           67    135    20   132    29    58    30    1    45    98  ...   
39          307   1960   395   887   252   379   243   29   287   541  ...   
40          349   1779   405  1003   238   566   381   31   279   760  ...   
41           27     47    11    25   133    72    84   18    68   103  ...   
42           94    178    24    54   256   758   502   23   254   945  ...   
43           72    134    13    36    67   151    73    4    88   285  ...   
44           83    320    61    80   189   432   187   13   128   632  ...   
999        1523    238    10    66   193   170   177   96   225   262  ...   
All        7210  10341  1703  5581  2971  5785  3630  574  3559  7958  ...   

Predicted     32    34     36    37  38     39      40   42  999     All  
True                                                                      
3              6     0     23     1   0    402     278    1    7    4517  
4              0     0      8     6   0     21     281    2    0     586  
5             28    12    198    35   0    310    3341    5    6    7438  
6              2     2     23     9   0    351    1191    0    1    2243  
7             37     6     64   250   0    359   11981    0    4   15374  
8            243   106    832   154   0   1364   11506    3    2   15244  
9            225     6    117    10   0    677    1578   55  108   11222  
12            11     1      9     3   0     27     970    0    0    1391  
14             0     0      0     0   0      4     

In [ ]:
from sklearn.metrics import classification_report   # classification report for train data
                                                            
print(classification_report(y_train,y_dtc_pred_train)) 
#[code snippet-77]

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

           3       0.50      0.80      0.61      4517
           4       0.00      0.00      0.00       586
           5       0.27      0.38      0.31      7438
           6       0.28      0.21      0.24      2243
           7       0.38      0.14      0.20     15374
           8       0.00      0.00      0.00     15244
           9       0.17      0.04      0.07     11222
          12       0.00      0.00      0.00      1391
          14       0.00      0.00      0.00        23
          15       0.26      0.32      0.29      4698
          18       0.19      0.36      0.25      1943
          19       0.20      0.15      0.17       795
          20       0.30      0.52      0.38      2075
          21       0.22      0.64      0.32      2677
          22       0.25      0.31      0.28      2362
          23       0.20      0.08      0.11       214
          24       0.32      0.45      0.38     11880
          25       0.42    

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
y_score_t=dtc.predict_proba(x1_train)
#[code snippet-78]

In [ ]:
roc_auc_score(y1_train,y_score_t,multi_class='ovr')   # calculating roc_auc score for train data
#[code snippet-79]

0.8303646201470584

**Hyper Parameter Tuning:-**

**1- By using manual tuning**

In [ ]:
x.columns     #[code snippet-80]

Index(['Weekday', 'DepartmentDescription'], dtype='object')

In [ ]:
  y.value_counts().head(10)      #[code snippet-81]

40    172705
39     94856
37     38645
38     29342
25     26642
7      23048
8      22630
36     21848
44     20093
42     19110
Name: TripType, dtype: int64

In [ ]:
y.shape       #[code snippet-82]

(636984,)

In [ ]:
dtc1=DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=5,    # forming instance 'dtc1' for DecisionTreeClassifier to perform manual tuning of the parameters
                       max_features=4, max_leaf_nodes=2,
                       min_impurity_decrease=0.01,                                  # in manual tuning we tune the parameters by hit and trail approach
                       min_samples_leaf=50, min_samples_split=10,
                       min_weight_fraction_leaf=0.0,
                       random_state=seed, splitter='random')
 #[code snippet-83]

In [ ]:
dtc1.fit(x1_train,y1_train)  # fitting DecisionTreeClassifier after tuning the parameters
 #[code snippet-84]

DecisionTreeClassifier(criterion='entropy', max_depth=5, max_features=4,
                       max_leaf_nodes=2, min_impurity_decrease=0.01,
                       min_samples_leaf=50, min_samples_split=10,
                       random_state=7, splitter='random')

In [ ]:
y_dtc_pred1=dtc1.predict(x1_test) # predicting  the values after training the tuned parameters
 #[code snippet-85]

In [ ]:
y_dtc_pred1
 #[code snippet-86]

array([40, 40, 40, ..., 40, 40, 40])

In [ ]:
y_dtc_pred_train1=dtc.predict(x1_train) # applying the DecisionTreeClassifier in train data and predicting it's values 
y_dtc_pred_train1
 #[code snippet-87]

array([40, 40, 40, ..., 25, 40,  5])

In [ ]:
confusion_matrix(y_test,y_dtc_pred1)  # confusion matrix for evaluating the model after manual tuning
pd.crosstab(y_test, y_dtc_pred1, rownames=['True'], colnames=['Predicted'],margins=True) 
 #[code snippet-88]

Predicted      40     All
True                     
3            2297    2297
4             303     303
5            3743    3743
6            1126    1126
7            7674    7674
8            7386    7386
9            5333    5333
12            701     701
14              6       6
15           2349    2349
18            988     988
19            388     388
20           1015    1015
21           1280    1280
22           1191    1191
23            109     109
24           5783    5783
25           8808    8808
26            831     831
27           1542    1542
28            859     859
29            675     675
30           1552    1552
31            601     601
32           4574    4574
33           3278    3278
34           1528    1528
35           4161    4161
36           7311    7311
37          12738   12738
38           9745    9745
39          31132   31132
40          56935   56935
41           1758    1758
42           6352    6352
43           2093    2093
44           6440    6440
999          5620    5620
All        210205  210205

In [ ]:
from sklearn.metrics import classification_report     # classification report for evaluating the model after manual tuning
                                                            
print(classification_report(y_test,y_dtc_pred1)) 
 #[code snippet-89]

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

           3       0.00      0.00      0.00      2297
           4       0.00      0.00      0.00       303
           5       0.00      0.00      0.00      3743
           6       0.00      0.00      0.00      1126
           7       0.00      0.00      0.00      7674
           8       0.00      0.00      0.00      7386
           9       0.00      0.00      0.00      5333
          12       0.00      0.00      0.00       701
          14       0.00      0.00      0.00         6
          15       0.00      0.00      0.00      2349
          18       0.00      0.00      0.00       988
          19       0.00      0.00      0.00       388
          20       0.00      0.00      0.00      1015
          21       0.00      0.00      0.00      1280
          22       0.00      0.00      0.00      1191
          23       0.00      0.00      0.00       109
          24       0.00      0.00      0.00      5783
          25       0.00    

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
y_score1=dtc1.predict_proba(x1_test)   # calculating the roc_auc score after tuning of parameter
roc_auc_score(y_test,y_score1,multi_class='ovr')
 #[code snippet-90]

0.524108222657591

In [ ]:
confusion_matrix(y_train,y_dtc_pred_train1)     # confusion matrix for train data
pd.crosstab(y_train, y_dtc_pred_train1, rownames=['True'], colnames=['Predicted'],margins=True)
#[code snippet-91]

Predicted     3      5     6     7     9    15    18   19    20    21  ...  \
True                                                                   ...   
3          3606     24     2     5     4    16     3    1    12    13  ...   
4             4    240     0     2     0     2     0    0     2     2  ...   
5            35   2791    18    55    41    51     9    0    30    93  ...   
6             8     31   474    32     5    17     3    0     3    16  ...   
7            42     65    30  2113    19    31    14    3    21    50  ...   
8             0    308    21   127    32    75    18    2    40    76  ...   
9            42    109     4    41   495   511   345   98   577   727  ...   
12            7     13     6     7     2    13    42    1     1    11  ...   
14            1      0     0     0     0     2     0    0     0     9  ...   
15           34     38     8    77    23  1506    23    2     4   214  ...   
18           21     28     1     8     9   109   704    1    11    63  ...   
19           29     22     3     2    12     4    13  116     5    32  ...   
20           32     32     4    14    25     5     9    3  1084    18  ...   
21           18     33     3     6    16    80    27    2     5  1716  ...   
22           62     28     5    22    38    28    12   93    15    29  ...   
23            8      2     0     1     5     1     0    0     0     7  ...   
24          121    164    17    58   155   118    54    8    44   339  ...   
25          142    222    24   120   134   158    90   10    48   248  ...   
26           20     34     2     8   182    10     7    0    23    41  ...   
27           24     49    11    16    14    37    18    0    16    37  ...   
28           28     17     4     6    10     8    14    3     8    21  ...   
29           25     32     4     5    13     7   295    2    18    17  ...   
30           34     59     5    23   198    30     6    0    19    76  ...   
31           21     18    12     5     3     7     5    3     5     9  ...   
32           64    121     7    57    48   128   109    1    29   101  ...   
33           33     64     6    37    25    22    20    1    23    41  ...   
34           21     49     3    22    13     8     4    1    36    22  ...   
35           48     58    14    86     8    33    14    2    16    53  ...   
36           78    713    22    61    61   130    47    2    54   165  ...   
37           80    186    54   282    24    52    48    4    36    86  ...   
38           67    135    20   132    29    58    30    1    45    98  ...   
39          307   1960   395   887   252   379   243   29   287   541  ...   
40          349   1779   405  1003   238   566   381   31   279   760  ...   
41           27     47    11    25   133    72    84   18    68   103  ...   
42           94    178    24    54   256   758   502   23   254   945  ...   
43           72    134    13    36    67   151    73    4    88   285  ...   
44           83    320    61    80   189   432   187   13   128   632  ...   
999        1523    238    10    66   193   170   177   96   225   262  ...   
All        7210  10341  1703  5581  2971  5785  3630  574  3559  7958  ...   

Predicted     32    34     36    37  38     39      40   42  999     All  
True                                                                      
3              6     0     23     1   0    402     278    1    7    4517  
4              0     0      8     6   0     21     281    2    0     586  
5             28    12    198    35   0    310    3341    5    6    7438  
6              2     2     23     9   0    351    1191    0    1    2243  
7             37     6     64   250   0    359   11981    0    4   15374  
8            243   106    832   154   0   1364   11506    3    2   15244  
9            225     6    117    10   0    677    1578   55  108   11222  
12            11     1      9     3   0     27     970    0    0    1391  
14             0     0      0     0   0      4     

In [ ]:
from sklearn.metrics import classification_report   # classification report for train data
                                                            
print(classification_report(y_train,y_dtc_pred_train1)) 
#[code snippet-92]

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

           3       0.50      0.80      0.61      4517
           4       0.00      0.00      0.00       586
           5       0.27      0.38      0.31      7438
           6       0.28      0.21      0.24      2243
           7       0.38      0.14      0.20     15374
           8       0.00      0.00      0.00     15244
           9       0.17      0.04      0.07     11222
          12       0.00      0.00      0.00      1391
          14       0.00      0.00      0.00        23
          15       0.26      0.32      0.29      4698
          18       0.19      0.36      0.25      1943
          19       0.20      0.15      0.17       795
          20       0.30      0.52      0.38      2075
          21       0.22      0.64      0.32      2677
          22       0.25      0.31      0.28      2362
          23       0.20      0.08      0.11       214
          24       0.32      0.45      0.38     11880
          25       0.42    

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


**2- Grid search cv**

In [ ]:
param_dist={
    'criterion':['gini','entropy'],'splitter':['best','random'],'max_depth':[2,4,6,8,10,12,None]
}     # parameters value dictionary for grid search cv
#[code snippet-93]

In [ ]:
from sklearn.model_selection import GridSearchCV      # importing grid search cv
grid=GridSearchCV(dtc, param_grid=param_dist,cv=10, n_jobs=-1)
#[code snippet-94]

In [ ]:
grid.fit(x1_train,y1_train)  # fitting GridSearchCV in train data
#[code snippet-95]

GridSearchCV(cv=10, estimator=DecisionTreeClassifier(), n_jobs=-1,
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [2, 4, 6, 8, 10, 12, None],
                         'splitter': ['best', 'random']})

In [ ]:
grid.best_estimator_  # shows the best parameters suitable for good tuning of decision tree model
#[code snippet-96]

DecisionTreeClassifier()

In [ ]:
grid.best_score_     # shows the best score obtained after tuning of hyper parameters
#[code snippet-97]

0.3358108068080302

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix,precision_score,f1_score

In [ ]:
p_criterion=['gini','entropy']
p_splitter=['best','random']   # implementing custom loop for hyperparameter tuning
p_max_depth=[50,100,150,200,250,300,350,400,450,500]
results=[]

for i in  p_criterion:
  for j in p_splitter:
    for k in p_max_depth:
      model= DecisionTreeClassifier(criterion=i, max_depth=k,random_state=seed, splitter=j)
      model.fit(x1_train,y1_train)
      preds=model.predict(x1_test)
      results.append({'Accuracy':round(accuracy_score(y_test,preds),5),'Precision':round(precision_score(y_test,preds,average='macro'),5),'F1-score':round(f1_score(y_test,preds,average='macro'),5),'p_criterion':i,'p_splitter':j,'p_max_depth':k})
results=pd.DataFrame(results)
results=results.sort_values(by='Accuracy',ascending=False)
print(results)

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Pr

    Accuracy  Precision  F1-score p_criterion p_splitter  p_max_depth
39   0.33652    0.20320   0.19533     entropy     random          500
36   0.33652    0.20320   0.19533     entropy     random          350
1    0.33652    0.20320   0.19533        gini       best          100
21   0.33652    0.20320   0.19533     entropy       best          100
22   0.33652    0.20320   0.19533     entropy       best          150
23   0.33652    0.20320   0.19533     entropy       best          200
24   0.33652    0.20320   0.19533     entropy       best          250
25   0.33652    0.20320   0.19533     entropy       best          300
26   0.33652    0.20320   0.19533     entropy       best          350
27   0.33652    0.20320   0.19533     entropy       best          400
28   0.33652    0.20320   0.19533     entropy       best          450
29   0.33652    0.20320   0.19533     entropy       best          500
37   0.33652    0.20320   0.19533     entropy     random          400
31   0.33652    0.20

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


**3- Randomized search cv**

In [ ]:
param_dist2={
    'criterion':['gini','entropy'],'splitter':['best','random'],'max_depth':[50,100,200,400,600,800,1000,1500,1800,2000,2200,2400,None],
    'min_samples_leaf':[200,400,600,800,1000,1200,1400,1600,2000,2400,2800,3000],
}   #'max_features':['sqrt','log2',None]     # parameters value dictionary for random search cv
#[code snippet-98]

In [ ]:
from sklearn.model_selection import RandomizedSearchCV   # importing the RandomizedSearchCV 
grid1=RandomizedSearchCV(estimator=dtc, param_distributions=param_dist2, cv=5, verbose=2, n_jobs=-1)
#[code snippet-99]

In [ ]:
grid1.fit(x1_train,y1_train)     # fitting RandomizedSearchCV in train data
#[code snippet-100]

Fitting 5 folds for each of 10 candidates, totalling 50 fits


RandomizedSearchCV(cv=5, estimator=DecisionTreeClassifier(), n_jobs=-1,
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'max_depth': [50, 100, 200, 400, 600,
                                                      800, 1000, 1500, 1800,
                                                      2000, 2200, 2400, None],
                                        'min_samples_leaf': [200, 400, 600, 800,
                                                             1000, 1200, 1400,
                                                             1600, 2000, 2400,
                                                             2800, 3000],
                                        'splitter': ['best', 'random']},
                   verbose=2)

In [ ]:
grid1.best_params_       # shows the best parameters suitable for good tuning of decision tree model
#[code snippet-101]

{'splitter': 'best',
 'min_samples_leaf': 800,
 'max_depth': 50,
 'criterion': 'gini'}

In [ ]:
grid1.best_score_      # shows the best score obtained after tuning of hyper parameters
#[code snippet-102]

0.33397847620631804

**Observations[code snippet:66-102]:-**

1- Decision Tree classifier is applied to the dataset for the classification of Trip Types.

2- Accuracy of 34%, test and train roc_auc score is 80% and 83%.

3- Manual tuning of Hyper parameter is done, with best score of 27%(since, manual tuning is tedious, so Grid search cv and Randomized search is also used). 

4- Grid search CV is used for Hyper parameter tuning, and a best score of 33.5% is obtained after tuning.

5- Randomized search CV is used for Hyper parameter tuning, and a best score of 33.3% is obtained after tuning.

6- Custom loop is used for hyper parameter tuning, and a best score of 33.6% is obtained after tuning. 

**Model 3- XgBoost classifier**

In [ ]:
from xgboost import XGBClassifier     # importing xgboost classifier
#[code snippet-103]

In [ ]:
xgb=XGBClassifier()     # initializing instance for xgboost classifier
#[code snippet-104]

In [ ]:
xgb.fit(x1_train,y1_train)    # fitting xgboost classifier in train data
#[code snippet-105]

XGBClassifier(objective='multi:softprob')

In [ ]:
y_xgb_pred=xgb.predict(x1_test)    # predicting the values by using test data 
#[code snippet-106]

In [ ]:
y_xgb_pred

array([40, 40,  5, ..., 34, 40, 40])

In [ ]:
y_xgb_pred_train=xgb.predict(x1_train)     # applying  model in train data to check robustness of model
y_xgb_pred_train
#[code snippet-108]

array([40, 40, 40, ..., 25, 40,  5])

In [ ]:
from sklearn.metrics import confusion_matrix      # importing confusion matrix
#[code snippet-109]

In [ ]:
confusion_matrix(y_test,y_xgb_pred)  # using confusion matrix for evaluating the performance of model using test data
pd.crosstab(y_test, y_xgb_pred, rownames=['True'], colnames=['Predicted'],margins=True) 
#[code snippet-110]

Predicted     3     5     6     7    15    18   19    20    21   22  ...  \
True                                                                 ...   
3          1833    12     2     8     1     2    4     8     5    5  ...   
4             1   166     0     2     1     0    0     1     0    0  ...   
5            15  1725    15    34    32     5    2     8    37    7  ...   
6             6     8   280    19    13     0    0     2     8    1  ...   
7            23    55    14  1244    13     3    0    10    11    1  ...   
8             0   178    19    72    26     8    2    16    39    3  ...   
9            19    54     2    25   207   112   73   305   244  136  ...   
12            0     9     2     8    11    17    1     3     4    2  ...   
14            1     0     0     0     1     0    0     0     1    0  ...   
15           16    20     7    42   765    10    0     1    67    5  ...   
18           11    13     2     6    68   275    1     6    19    1  ...   
19           18     5     2     2     3     0   74     3    12   89  ...   
20           16    28     3     8     4     4    0   544     9    2  ...   
21            8    17     0     4    44    13    0     0   659    2  ...   
22           33    26     2     8     9     4   81    11    12  261  ...   
23            8     5     0     1     0     1    1     0     0   17  ...   
24           67    89    10    41    60    18    8    26   110   12  ...   
25           56   178    21    72    82    28    4    27    88   11  ...   
26            9    19     4     7     4     1    1    16     9    0  ...   
27           10    24     6     5    20     4    0     5     9    1  ...   
28           12     8     2     6    10     4    0     7     6    2  ...   
29           12    15     2     5     2   101    1    12     4    1  ...   
30           22    18     4     7    17     6    3     6     9    9  ...   
31            6    15     5     5     2     0    4     4     3    6  ...   
32           27    68     7    29    70    41    0    14    34    7  ...   
33           15    39     7    17    15     4    2     8    20    5  ...   
34           13    30     4    12     5     3    1    14     6    1  ...   
35           20    33    10    42    11     9    0    11    25    1  ...   
36           45   466    14    26    70    24    0    34    41   16  ...   
37           41   110    64   224    34    26    1    22    15   10  ...   
38           41    96    15    76    32    15    2    11    34    2  ...   
39          132  1257   246   516   198   115   19   145   183   59  ...   
40          174  1134   293   646   261   125   25   129   244   63  ...   
41           12    29     9     7    47    32    4    36    28   14  ...   
42           25   127    14    38   366   166   14   122   315   62  ...   
43           20   100     5    36    69    32    5    41    90    8  ...   
44           46   235    39    41   180    64    8    81   184   33  ...   
999         767   150    11    68    67    77   79   117    72  123  ...   
All        3580  6561  1142  3409  2820  1349  420  1806  2656  978  ...   

Predicted    28    30   31    32    34    36    37    39      40     All  
True                                                                      
3             4     8    2     6     2     4     2   200     152    2297  
4             0     2    0     0     1     3     3     8     109     303  
5            13    10    3    10    31   109    21   115    1395    3743  
6             4     5    1     2     5     8     7    49     677    1126  
7            12    10    3    18    14    31   114   148    5846    7674  
8            13    26    4   132   133   412    85   541    5576    7386  
9           212   270   79   129     9    43    10   280    1446    5333  
12            6     5    0     1     3    10     2    10     496     701  
14            0     0    0     0     1     0     0     0       2       6  
15            6    15    1    37     2    20    21    96  

In [ ]:
from sklearn.metrics import classification_report     # classification report for evaluating the model using test data
                                                            
print(classification_report(y_test,y_xgb_pred)) 
#[code snippet-111]

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

           3       0.51      0.80      0.62      2297
           4       0.00      0.00      0.00       303
           5       0.26      0.46      0.33      3743
           6       0.25      0.25      0.25      1126
           7       0.36      0.16      0.22      7674
           8       0.00      0.00      0.00      7386
           9       0.00      0.00      0.00      5333
          12       0.00      0.00      0.00       701
          14       0.00      0.00      0.00         6
          15       0.27      0.33      0.30      2349
          18       0.20      0.28      0.24       988
          19       0.18      0.19      0.18       388
          20       0.30      0.54      0.39      1015
          21       0.25      0.51      0.33      1280
          22       0.27      0.22      0.24      1191
          23       0.00      0.00      0.00       109
          24       0.31      0.43      0.36      5783
          25       0.42    

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
confusion_matrix(y_train,y_xgb_pred_train)  # confusion matrix for evaluating the model using train data
pd.crosstab(y_train,y_xgb_pred_train, rownames=['True'], colnames=['Predicted'],margins=True) 
#[code snippet-112]

Predicted     3      5     6     7    15    18   19    20    21    22  ...  \
True                                                                   ...   
3          3606     30     4     5    16     1    5    12    10     6  ...   
4             4    309     0     7     2     0    0     2     2     0  ...   
5            35   3374    20    63    51     7    1    30    66    18  ...   
6             8     40   619    37    17     1    0     3    12     2  ...   
7            42     84    45  2422    31    10    5    21    36    10  ...   
8             0    379    26   145    75    15    4    40    60     5  ...   
9            42    136     4    45   511   264  143   577   541   327  ...   
12            7     19     8     9    13    31    1     1     2     3  ...   
14            1      0     0     0     2     0    0     0     6     2  ...   
15           34     44    11    85  1506    21    4     4   148    10  ...   
18           21     32     2     9   109   539    1    11    45    10  ...   
19           29     25     4     3     4    12  164     5    26   169  ...   
20           32     45     6    15     5     7    3  1084    11     2  ...   
21           18     40     3     8    80    26    2     5  1369     4  ...   
22           62     37     8    26    28    11  145    15    20   492  ...   
23            8      3     0     1     1     0    1     0     7    27  ...   
24          121    215    22    75   118    43   11    44   237    26  ...   
25          142    309    35   138   158    71   13    48   151    32  ...   
26           20     37     2    10    10     3    2    23    23     2  ...   
27           24     68    14    20    37    14    0    16    24     2  ...   
28           28     26     7     6     8    10    3     8    15     5  ...   
29           25     40     5     5     7   222    4    18     9     3  ...   
30           34     80     6    28    30     4    2    19    49     5  ...   
31           21     25    12     5     7     3    5     5     5    13  ...   
32           64    140     9    63   128    74    1    29    75    21  ...   
33           33     71    11    44    22    16    1    23    29     6  ...   
34           21     72     9    24     8     1    1    36    18     3  ...   
35           48     75    17    96    33    14    2    16    40     4  ...   
36           78    891    24    68   130    36    2    54   104    20  ...   
37           80    226    85   355    52    31    4    36    55    26  ...   
38           67    175    34   175    58    20    1    45    72     6  ...   
39          307   2472   534  1058   379   172   37   287   369   144  ...   
40          349   2413   550  1329   566   247   38   279   493   144  ...   
41           27     66    13    33    72    63   26    68    54    57  ...   
42           94    222    28    65   758   388   35   254   670   165  ...   
43           72    174    14    43   151    60    5    88   204    14  ...   
44           83    453    75   105   432   145   16   128   423    74  ...   
999        1523    288    16    76   170   122  144   225   193   303  ...   
All        7210  13135  2282  6701  5785  2704  832  3559  5673  2162  ...   

Predicted    28    30    31     32    34     36    37     39      40     All  
True                                                                          
3            12     7     1      6     5     23     1    391     300    4517  
4             2     0     0      0     0      8     6     18     214     586  
5            25    33     8     28    56    198    35    254    2878    7438  
6             5     5     0      2    14     23     9     87    1306    2243  
7            17    24     8     37    19     64   250    284   11769   15374  
8            18    38     4    243   314    832   154   1159   11502   15244  
9           438   593   201    225    16    117    10    601    3018   11222  
12           11     8     0     11     4      9     3     23    1010    1391  
14         

In [ ]:
from sklearn.metrics import classification_report     # classification report for evaluating the model using train data
                                                            
print(classification_report(y_train,y_xgb_pred_train)) 
#[code snippet-113]

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

           3       0.50      0.80      0.61      4517
           4       0.00      0.00      0.00       586
           5       0.26      0.45      0.33      7438
           6       0.27      0.28      0.27      2243
           7       0.36      0.16      0.22     15374
           8       0.00      0.00      0.00     15244
           9       0.00      0.00      0.00     11222
          12       0.00      0.00      0.00      1391
          14       0.00      0.00      0.00        23
          15       0.26      0.32      0.29      4698
          18       0.20      0.28      0.23      1943
          19       0.20      0.21      0.20       795
          20       0.30      0.52      0.38      2075
          21       0.24      0.51      0.33      2677
          22       0.23      0.21      0.22      2362
          23       0.00      0.00      0.00       214
          24       0.33      0.44      0.38     11880
          25       0.42    

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
from sklearn.metrics import roc_auc_score      # calculating the roc_auc score 
y_xgb_score=xgb.predict_proba(x1_test)
#[code snippet-114]

In [ ]:
roc_auc_score(y_test,y_xgb_score,multi_class='ovr')   # roc_auc score using test data
#[code snippet-115]

0.7866433360515903

In [ ]:
y_xgb_score_train=xgb.predict_proba(x1_train)    # roc_auc score using train data
#[code snippet-116]

In [ ]:
roc_auc_score(y1_train,y_xgb_score_train,multi_class='ovr')   #[code snippet-117]

0.7928994554065774

**Observations[code snippet:103-123]:-**

1- Xgboost classifier is applied to the dataset for the classification of Trip Types.

2- Accuracy of 33%, test and train roc_auc scores is 78% and 79% are obtained after applying Xgboost classifier.

